### PR TITLE
Sort types for display

### LINKED ISSUE #32224: [Feature request] sort string literal in .d.ts
## Suggestion

sort string literal in .d.ts

## Use Cases

avoid string literal random change order when source code didn't change
and make code history is more easy read

in this issue thread `string literal` is mean => `"a" | "c" | "b"`

this not only happen on `Record`, it has chance happen at all type is auto create by typescript emit with `string literal`

## Examples

> `"a" | "c" | "b"` is create by typesctipt emit and some time will random change order

- https://github.com/bluelovers/cjk-convert/blob/1a0ba9f85e0bf7fe916158cc14a4ef87507106d9/lib/zh/table/table.ts#L1619-L1635
- https://github.com/microsoft/TypeScript/issues/30328
  - https://github.com/bluelovers/novel-travis-test/commit/3b88745e187736797d9a30d4f4bbd927b0c6dd30#diff-3006c96d8657ccca5b9b3618a256314c
  - https://github.com/bluelovers/novel-travis-test/commit/d33c4a61602a8536d1414d36e027eb99845104f1#diff-3006c96d8657ccca5b9b3618a256314c
  - https://github.com/bluelovers/novel-travis-test/commit/2d749195c39ef386e6c4bad1ae6e19d007c03e93#diff-3006c96d8657ccca5b9b3618a256314c

### current .d.ts output

---

```ts
export declare const table_plus: Record<"a" | "c" | "b">
// "a" | "c" | "b" is random change order
```

### in this request

> sort ( use simple array.sort() ) it when output emit at `.d.ts`

```ts
export declare const table_plus: Record<"a" | "b" | "c">
```

## Checklist

My suggestion meets these guidelines:

* [x] This wouldn't be a breaking change in existing TypeScript/JavaScript code
* [x] This wouldn't change the runtime behavior of existing JavaScript code
* [ ] This could be implemented without emitting different JS based on the types of the expressions
* [x] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, etc.)
* [ ] This feature would agree with the rest of [TypeScript's Design Goals](https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals).

### PR DESCRIPTION
**Pull request by sandersn**
Pushing to a PR for performance testing. This initial cut looks extremely slow.

Fixes #32224

### CODE REVIEW

**Comment by sandersn** [comment]
@typescript-bot perf test this

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -365,6 +365,93 @@ namespace ts {
 
         const argumentsSymbol = createSymbol(SymbolFlags.Property, "arguments" as __String);
         const requireSymbol = createSymbol(SymbolFlags.Property, "require" as __String);
+        const fail = (name: string) => () => Debug.fail("Unit kind expected for " + name)
+        // boolean, bigint, number, string, symbol [literal], boolean, bigint, number, string, symbol, object, union (recursive?!), ..., unknown, never, null, void, undefined, any
+        const ordre: Array<[TypeFlags, (t1: Type, t2: Type) => Comparison]> = [
```
Comment (a-tarasyuk) [inline src/compiler/checker.ts:370]: order

Comment (sandersn) [reply src/compiler/checker.ts:370]: This is a draft, so this is just the name that fell out of my brain and it happened to be in French. (I often use short, intentionally bad names so I'll be forced to change them later.)

**Comment by DanielRosenwasser** [comment]
Do we have declaration emit captured in the perf tests?

https://github.com/microsoft/TypeScript/issues/44033

File: src/compiler/checker.ts (hunk 1 above)
Comment (a-tarasyuk) [reply src/compiler/checker.ts:370]: Got it :)

**Comment by sandersn** [comment]
@amcasey seemed to think not, so on his recommendation I have been testing by compiling tinymce instead. However, it has around 1000 unions, which may not be enough to detect a worst-case slowdown. Maybe I should find a gnarly DT package, read it, then print it out and see how long printing it takes.

**Comment by sandersn** [comment]
Notes from discussion in our team meeting:

1. typeToString already puts null/undefined at end, need to dedupe this code
2. enums should probably maintain order
3. true | false collapses to boolean
4. should maintain order of properties dumped via keyof
5. decide what to do with origin types from denormalised unions

**Comment by sandersn** [comment]
Well, this was an interesting experiment, but after finishing the sort code, I found the best performance test is our type baselines. I just ran `time gulp runtests-parallel` and looked at the amount of CPU time spent. It's 10% slower even skipping the less-important-to-sort types. So I don't think this is a viable approach. Probably the only thing that would work is to sort during union creation.

**Pull request closed without merging**